### PR TITLE
Use predicates which are pushed down to Parquet reader in CheckpointEntryIterator

### PR DESCRIPTION
**Pull request by jkylling**
## Description
Fixes https://github.com/trinodb/trino/issues/17405

Instead of a not null filter on the entry type column, which is a complex type for which null filters are NOOPs, we use a not null filter on a primitive required field of the entry type. This enables data skipping in the Parquet reader.

## Additional context and related issues

Please see https://github.com/trinodb/trino/issues/17405

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Delta
* Improve performance of reading from tables with a large number of checkpoint entries. ({issue}`17405`)
```

### CODE REVIEW

**Comment by jkylling** [comment]
It is tricky to test the feature directly without further modifications.

With custom tests I've verified that the feature works by:
1. Enabling debug logs and seeing that there are much fewer `Building add entry from %s pagePosition %d"` without a corresponding `Block %s has %s fields` when loading a checkpoint.
2. Timing the time it takes to load a 300 MB checkpoint (from ~40 seconds to ~1 second), and the time it takes to load an individual checkpoint part (from ~500 ms to < 2 ms).

I tried using a TrackingFileSystem in https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java, but the number of I/O operations were unchanged with or without the fix, on both the existing `TEST_CHECKPOINT`, and the custom checkpoint part.
A possible approach is to use a logger which counts the number of times it is called, and verify this number.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 1]
```diff
@@ -225,6 +229,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
         return new DeltaLakeColumnHandle(entryType.getColumnName(), type, OptionalInt.empty(), entryType.getColumnName(), type, REGULAR, Optional.empty());
     }
 
+    /**
+     * Constructs a TupleDomain which filters on a specific required column of the EntryType being not null. This speeds
+     * up reading of large checkpoint files where the majority of the entries are remove tombstones, and most row groups
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:234]: > This speeds up reading of large checkpoint files where the majority of the entries are remove tombstones, and most row groups can be skipped by reading the parquet metadata.

This is much too verbose.


Try

> Constructs a TupleDomain which filters on a specific required primitive sub-column of the `EntryType` being not null for effectively pushing down the predicate to the Parquet reader.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java [hunk 2]
```diff
@@ -105,6 +105,24 @@ public void testReadMetadataEntry()
                                 1579190100722L));
     }
 
+    @Test
+    public void testReadProtocolEntries()
+            throws Exception
+    {
+        URI checkpointUri = getResource(TEST_CHECKPOINT).toURI();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:112]: This test is not effective to showcase your fix.

Let's try a different approach:

- expose `getCompletedBytes()` `VisibleForTesting` in `CheckpointEntryIterator`
- use a simplified version of  `io.trino.plugin.deltalake.transactionlog.checkpoint.TestCheckpointWriter#testCheckpointWriteReadJsonRoundtrip` for creating a checkpoint on the fly with one entry of addfile and a lot of removefile entries
- check after getting the addfile entries from the checkpoint entry iterator whether the number of bytes read from the parquet checkpoint file (through previously added `getCompletedBytes()` method) is indeed much smaller than the number of bytes read while reading the remove file entries.
I'm assuming that a non existent entry type in the checkpoint should also imply reading `0` bytes from the parquet file. To be verified.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 3]
```diff
@@ -225,6 +229,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+                type = BigintType.BIGINT;
+                break;
+            case COMMIT:
+                field = "version";
+                type = BigintType.BIGINT;
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported Delta Lake checkpoint entry type: " + entryType);
+        }
+        HiveColumnHandle handle = new DeltaLakeColumnHandle(
```
[the remark this answers is not in the document]
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java]: Thank you, this simplified the code

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 1 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java]: Indeed, this is better

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java (hunk 2 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:120]: Thank you for the suggestion, I've added a test following this recipe. 

The completedBytes of the Parquet reader includes the bytes used to read the Parquet metadata, hence `txnEntryIterator.getCompletedBytes() != 0`. This also means that we must use a fairly large number of remove entries with random values, such that the size of the remove entries is large relative to the Parquet metadata size.

The test is potentially a bit sensitive to changes to the Parquet writer and reader (e.g., if the writer is modified to write everything as a single row group, or the header becomes very large). For the chosen values the metadata is ~10% of the file size, so the test should not be flaky.

The CheckpointWriter uses the default values for the ParquetWriter. Potentially, these values should be tuned, as the checkpoint files likely have a different access patterns than what is common for Parquet files. It would be beneficial if all non-remove entries are in as few row groups as possible.

`testReadProtocolEntries` was added to have a read test for protocol entries, and to manually test that we had coverage for invalid specific column types. I can remove it if necessary?

Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:120]: Take a look at  `CheckpointWriter` 

https://github.com/trinodb/trino/blob/962c0cc7bc4fcd3f20a8fa0765c651ee954ba37c/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java#L127

Now, if you make a slight change on `CheckpointWriter` to provide a  `ParquetWriterOptions` parameter in the constructor, you could customise the size of the row groups and you would probably be able to showcase relatively easy the row group pruning:

```
        ParquetWriterOptions parquetWriterOptions = ParquetWriterOptions.builder()
                .setMaxBlockSize(DataSize.ofBytes(100L))
                .setMaxPageSize(DataSize.ofBytes(100L))
                .build();
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 4]
```diff
@@ -225,6 +231,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+            default:
+                throw new IllegalArgumentException("Unsupported Delta Lake checkpoint entry type: " + entryType);
+        }
+        HiveColumnHandle handle = new HiveColumnHandle(
+                column.getBaseColumnName(),
+                column.getBaseHiveColumnIndex(),
+                column.getBaseHiveType(),
+                column.getBaseType(),
+                Optional.of(new HiveColumnProjectionInfo(
+                        ImmutableList.of(-1), // hiveColumnIndex; we provide fake value because we always find columns by name
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:276]: nit: Use `0` - this is what i see in `io.trino.plugin.deltalake.DeltaLakeColumnHandle#toHiveColumnHandle`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java [hunk 5]
```diff
@@ -91,6 +92,13 @@ public CheckpointWriter(TypeManager typeManager, CheckpointSchemaManager checkpo
 
     public void write(CheckpointEntries entries, TrinoOutputFile outputFile)
             throws IOException
+    {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java:95]: Extract this addition to a separate commit.

**Review by findinpath — COMMENTED** [review]
LGTM 

Pls restructure the PR to have 2 commits:
- Add method to customise parquet writer options in `CheckpointWriter`
- Push down predicates to Parquet in `CheckpointEntryIterator`

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 6]
```diff
@@ -225,6 +231,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+        HiveColumnHandle handle = new HiveColumnHandle(
+                column.getBaseColumnName(),
+                column.getBaseHiveColumnIndex(),
+                column.getBaseHiveType(),
+                column.getBaseType(),
+                Optional.of(new HiveColumnProjectionInfo(
+                        ImmutableList.of(0), // hiveColumnIndex; we provide fake value because we always find columns by name
+                        ImmutableList.of(field),
+                        HiveType.toHiveType(type),
+                                type)),
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:279]: Fix the indentation on this line

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 7]
```diff
@@ -225,6 +231,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+    /**
+     * Constructs a TupleDomain which filters on a specific required primitive sub-column of the EntryType being
+     * not null for effectively pushing down the predicate to the Parquet reader.
+     */
+    private TupleDomain<HiveColumnHandle> buildTupleDomainColumnHandle(EntryType entryType, HiveColumnHandle column)
+    {
+        String field;
+        Type type;
+        switch (entryType) {
+            case TRANSACTION:
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:243]: You could combine some of these cases.
```suggestion
            case COMMIT:
            case TRANSACTION:
```

Same deal with ADD and REMOVE

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 8]
```diff
@@ -225,6 +231,57 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+                field = "version";
+                type = BigintType.BIGINT;
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:245]: Usually static imports of these is preferred.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java [hunk 9]
```diff
@@ -57,6 +66,7 @@
 {
     private static final String TEST_CHECKPOINT = "databricks/person/_delta_log/00000000000000000010.checkpoint.parquet";
 
+    private final TypeManager typeManager = TESTING_TYPE_MANAGER;
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:69]: You can just use TESTING_TYPE_MANAGER directly, without the extra variable definition here

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java [hunk 10]
```diff
@@ -207,6 +235,88 @@ public void testReadAllEntries()
...
+        writer.write(
+                entries,
+                createOutputFile(targetPath),
+                ParquetWriterOptions.builder() // approximately 2 rows per row group
+                        .setMaxBlockSize(DataSize.ofBytes(64L))
+                        .setMaxPageSize(DataSize.ofBytes(64L))
+                        .build());
+
+        CheckpointEntryIterator addEntryIterator =
+                createCheckpointEntryIterator(URI.create(targetPath), ImmutableSet.of(ADD), Optional.of(readMetadataEntry(URI.create(targetPath))));
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:306]: Very nitpicky, but I'd format this more like

```suggestion
        CheckpointEntryIterator addEntryIterator = createCheckpointEntryIterator(
            URI.create(targetPath),
            ImmutableSet.of(ADD), 
            Optional.of(readMetadataEntry(URI.create(targetPath))));
```

**Review by alexjo2144 — COMMENTED** [review]
Thanks for the fix, just a couple small comments.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 6 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java]: Thanks. Looks like the checkstyle does not enforce all of the Intellij Airlift style which I just recently installed.

**Comment by jkylling** [comment]
@findinpath Can this PR be merged?

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java [hunk 11]
```diff
@@ -207,6 +233,91 @@ public void testReadAllEntries()
...
+                ParquetWriterOptions.builder() // approximately 2 rows per row group
+                        .setMaxBlockSize(DataSize.ofBytes(64L))
+                        .setMaxPageSize(DataSize.ofBytes(64L))
+                        .build());
+
+        CheckpointEntryIterator addEntryIterator =
+                createCheckpointEntryIterator(
+                        URI.create(targetPath),
+                        ImmutableSet.of(ADD),
+                        Optional.of(readMetadataEntry(URI.create(targetPath))));
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestCheckpointEntryIterator.java:307]:
```suggestion
                Optional.of(metadataEntry));
```

**Comment by findinpath** [comment]
@jkylling pls address the open comments

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 12]
```diff
@@ -225,6 +231,51 @@ private DeltaLakeColumnHandle buildColumnHandle(EntryType entryType, CheckpointS
...
+     * not null for effectively pushing down the predicate to the Parquet reader.
+     */
+    private TupleDomain<HiveColumnHandle> buildTupleDomainColumnHandle(EntryType entryType, HiveColumnHandle column)
+    {
+        String field;
+        Type type;
+        switch (entryType) {
+            case COMMIT:
+            case TRANSACTION:
+                field = "version";
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:245]: How do we know which fields exist in the column and which ones can be non-null ? If it's defined somewhere in the code or in the delta spec, it would help to have a code comment pointing to it

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java [hunk 13]
```diff
@@ -627,6 +678,12 @@ private void fillNextEntries()
         }
     }
 
+    @VisibleForTesting
+    long getCompletedBytes()
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:682]: Can we use `getCompletedPositions` instead for testing ? It should give you the number of input rows in the parquet reader and should be a more stable number than the bytes read.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java [hunk 14]
```diff
@@ -91,6 +92,13 @@ public CheckpointWriter(TypeManager typeManager, CheckpointSchemaManager checkpo
...
+        write(entries, outputFile, ParquetWriterOptions.builder().build());
+    }
+
+    @VisibleForTesting
+    void write(CheckpointEntries entries, TrinoOutputFile outputFile, ParquetWriterOptions parquetWriterOptions)
```
Comment (raunaqmorarka) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java:100]: `ParquetWriterOptions` should probably be an argument to the constructor of `CheckpointWriter`. It seems safer to use the same options to write all files rather than allowing different config for every file.

**Comment by findinpath** [comment]
@jkylling  pls address the open comments, do a push of them and then do another push for rebasing on top of `master` to resolve the conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 12 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:245]: It's part of the delta spec, and we also enforce it when constructing action entries. For most of the entry types this is listed as optional/required in the description tables in the delta spec, except for the Transaction and Protocol actions. I've opened https://github.com/delta-io/delta/pull/1890 to make it clearer.

Actually, it looks like we are too strict when reading transaction entries, and require that the `lastUpdated` field is set, even though it is optional in the spec: https://github.com/trinodb/trino/blob/e920b4b1e1f091ed9d2ceec82e50a83996599544/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java#L521. It should not cause any issues, as I believe we don't use Transaction entries anywhere, but it would be good to fix.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java (hunk 13 above)
Comment (jkylling) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointEntryIterator.java:682]: This is much cleaner. We can now assert exactly how many positions will be read per type.

**Review by raunaqmorarka — APPROVED** [review]
